### PR TITLE
Implements reproduction of Figure 5. RNN Bifurcations

### PR DESCRIPTION
**Pull request by evaisd**
# Module helpers.py with the functions that do the analysis of the RNN.
# Script plot_fig5.py (self explanatory).
# Some Other additions/fixes:
1. Fixed single/double comma issue in train.py. 
2. Fixed the ZRNN to run without the Euler delta term during inference.
3. Added continuation_phase parameter to generate_stimuli that makes it generate null I_stim.
4. Some fixes in the ZRNN code - fixed devices compatibility issues and parameters names.

### CODE REVIEW

File: zrnn/datasets.py [hunk 1]
```diff
@@ -2,15 +2,16 @@
 import torch
 from torch.utils.data import Dataset
 
-def generate_stimuli(period, T_onset, duration=2.0, dt=0.001, one_dur=0.01):
+def generate_stimuli(period, T_onset, duration=2.0, dt=0.001, one_dur=0.01, continuation_phase: bool = False):
```
Comment (iranroman) [inline zrnn/datasets.py]: I'd say we should change the name of this keyword argument from `continuation_phase` to `I_stim_without_clicks` and flip the boolean logic.

**Comment by iranroman** [comment]
Thanks for this awesome work @evaisd . Please also make changes to the README to include the steps necessary to run this analysis.

File: zrnn/helpers.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+
+# def get_vector_field(model: ZemlianovaRNN,
+#                      context_cue_amplitude: float,
+#                      initial_neuron_activity: torch.Tensor,
+#                      trained_pca: PCA,
+#                      bounds: (float, float),
+#                      time_steps_ms: int = 500,
+#                      grid_length: int = 16,
+#                      at_time: int = None) -> dict:
```
Comment (iranroman) [inline zrnn/helpers.py]: if this stuff in comments is not needed, let's delete it please.

File: zrnn/helpers.py
Comment (iranroman) [inline zrnn/helpers.py:1]: can we rename this as `utils.py` please?

File: zrnn/helpers.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
+
+from pathlib import Path
+import numpy as np
+import torch
+import pandas as pd
```
Comment (iranroman) [inline zrnn/helpers.py]: please add pandas to `requirements.txt`

**Comment by iranroman** [comment]
This looks great overall. I left some minor comments, I might add more whenever I am able to run this to better understand the mechanics.  

Let's definitively add instructions to the README. Right now, for example, I just tried cloning the repo and running `python plot_fig5.py` but got this cryptic error. I'm sure I'm not following some appropriate step, but I have no way of knowing. 

```
ZemlianovaRNN % python plot_fig5.py
Traceback (most recent call last):
  File "/Users/iranroman/models/ZemlianovaRNN/plot_fig5.py", line 135, in <module>
    Fire(main)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/lib/python3.12/site-packages/fire/core.py", line 135, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniconda/base/envs/lib/python3.12/site-packages/fire/core.py", line 468, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
                                ^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniconda/base/envs/lib/python3.12/site-packages/fire/core.py", line 684, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/iranroman/models/ZemlianovaRNN/plot_fig5.py", line 130, in main
    _plot_a(span=span, grid_length=grid_length)
  File "/Users/iranroman/models/ZemlianovaRNN/plot_fig5.py", line 44, in _plot_a
    im = _gen_single_plot(data, ax, span, grid_length, center_around_end_point=True)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/iranroman/models/ZemlianovaRNN/plot_fig5.py", line 26, in _gen_single_plot
    end_point = (data['trajectory0'][-1], data['trajectory1'][-1])
                 ~~~~^^^^^^^^^^^^^^^
KeyError: 'trajectory0'
```

**Comment by iranroman** [comment]
@kzemlianova please feel free to have a look and leave a review

**Comment by iranroman** [comment]
@IamMarcIvanov (i.e. Danish) you too may leave a review

**Comment by evaisd** [comment]
@iranroman I've got the issues fixed. Now it should run smoothly. It failed when you tried executing it because I left a flag I used to avoid repeating some steps during development. I updated the readme as well.

File: README_figs/fig_5b.png
Comment (iranroman) [inline README_figs/fig_5b.png:1]: Wow, these figures are truly cool. 

One question: in this figure, is there a specific reason why you used timepoints that are different to those in the paper?

File: plot_fig5.py
Comment (iranroman) [inline plot_fig5.py:1]: can we add a progress bar or some sort of printing statement to let the user know what stage of the analysis we are at?

Also, running this I could not reproduce the qualitative features of the figures you provide. Granted, I was using a set of model weights I already had saved somewhere. Let me re-try from scratch running all pieces of code as you specify in this PR from the top of the README.

File: zrnn/helpers.py
Comment (iranroman) [reply zrnn/helpers.py:1]: @evaisd what do you think about this suggestion?

**Review by iranroman — CHANGES_REQUESTED** [review]
Thanks! I left a few more comments. In my view, this is almost there.

File: plot_fig5.py
Comment (iranroman) [reply plot_fig5.py:1]: Nevermind, I could! I may have been using an old set of `best_model.pth` weights.

File: zrnn/helpers.py (hunk 2 above)
Comment (evaisd) [reply zrnn/helpers.py]: Fixed in the more recent push.

File: zrnn/helpers.py (hunk 3 above)
Comment (evaisd) [reply zrnn/helpers.py]: Added in the more recent push.

File: plot_fig5.py
Comment (evaisd) [reply plot_fig5.py:1]: Added progress statements.

File: README_figs/fig_5b.png
Comment (evaisd) [reply README_figs/fig_5b.png:1]: It's a combination of me mistaking the initial condition in this plot (in the paper they're not on the balance manifold as I did but on 0), which I have fixed, and the fact that in our simulation the flow seems to be a bit slower. In any case, I have reset the timepoints to be the same as in the paper.

File: zrnn/helpers.py
Comment (evaisd) [reply zrnn/helpers.py:1]: I don't mind, will rename it utils.

**Review by iranroman — APPROVED** [review]
Wow. Truly amazing work @evaisd I'm in awe. We are ready to merge. 

@kzemlianova and @IamMarcIvanov I encourage you to have a look.

**Pull request merged**